### PR TITLE
better diagnostics for `impl<..> impl Trait for Type`

### LINKED ISSUE #109963: Want better error for  impl<some generics> impl Trait for Type {...}
### Code

```Rust
use std::panic::UnwindSafe;

struct S<T>(T);

impl<T: UnwindSafe> impl UnwindSafe for T {}

// We're using UnwindSafe as an example trait
```

### Current output

```Shell
error: expected a trait, found type
 --> src/lib.rs:5:21
  |
5 | impl<T: UnwindSafe> impl UnwindSafe for T {}
  |                     ^^^^^^^^^^^^^^^
```

### Desired output

```Shell
error: expected a trait, found type
 --> src/lib.rs:5:21
  |
5 | impl<T: UnwindSafe> impl UnwindSafe for T {}
  |                     ^^^^^^^^^^^^^^^

 --> src/lib.rs:5:21
  |
5 | impl<T: UnwindSafe> impl UnwindSafe for T {}
  | ^^^^                ^^^^
note: second occurrence of keyword `impl` is probably a mistake
```

### Rationale and extra context

I made this mistake and didn't spot that the `^^^^` were pointing at the whole of `impl Trait` rather than just `Trait` and spent several minutes thinking "but `Trait` *is* a trait!".

Probably my "desired output" could be improved.  Maybe *replacing* rather than *supplementing* the "expected trait, not type" would be better.

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by y21**
Fixes #109963

### CODE REVIEW

File: compiler/rustc_parse/src/parser/item.rs [hunk 1]
```diff
@@ -603,10 +603,25 @@ impl<'a> Parser<'a> {
...
-                            span: ty_first.span,
-                        });
+                    other => {
+                        // Suggest removing extra `impl` keyword:
+                        // `impl<T: Default> impl Default for Wrapper<T>`
+                        //                   ^^^^
+                        if let TyKind::ImplTrait(..) = other {
+                            let lo_span = ty_first.span.shrink_to_lo();
+                            let hi_span =
+                                lo_span.with_lo(lo_span.lo() + BytePos("impl".len() as u32));
```
Comment (y21) [inline compiler/rustc_parse/src/parser/item.rs:613]: This is ugly but I don't know if there's a better way to get a span of the `impl` bit of the `impl Trait` type

Comment (y21) [reply compiler/rustc_parse/src/parser/item.rs]: I decided to rework this entirely by not using the parsed `impl Trait` type at all and instead check if the `impl` keyword is present after generics. No need to hack the spans together that way

File: compiler/rustc_parse/src/parser/item.rs [hunk 2]
```diff
@@ -604,9 +609,17 @@ impl<'a> Parser<'a> {
                     // This notably includes paths passed through `ty` macro fragments (#46438).
                     TyKind::Path(None, path) => path,
                     _ => {
-                        self.sess.emit_err(errors::ExpectedTraitInTraitImplFoundType {
-                            span: ty_first.span,
-                        });
+                        if self.may_recover() && let Some(extra_impl_kw) = extra_impl_kw {
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/item.rs:612]: I actually liked the previous approach. You probably need to check that the `impl Trait` only has one bound (isn't `impl A + B`) and you can probably use the span from the beginning of the ty to the beginning of the first bound for the removal span.

**Comment by compiler-errors** [comment]
Please squash this into one commit. Then I can approve, thanks

**Pull request merged**